File: 3609.minimum-moves-to-reach-target-in-grid.java
#
# @lc app=leetcode id=3609 lang=java
#
# [3609] Minimum Moves to Reach Target in Grid
#
# @lc code=start
class Solution {
    public int minMoves(int sx, int sy, int tx, int ty) {
        int moves = 0;
        while (tx > sx || ty > sy) {
            if (tx < sx || ty < sy) return -1;
            if (tx > ty) {
                int m = Math.max(tx, ty);
                int prev_tx = Math.max(sx, ty);
                // Reduce as much as possible in one step
                int jump = (tx - sx - 1) / m + 1;
                if (ty == sy) jump = 1;
                tx -= m * jump;
                moves += jump;
            } else if (ty > tx) {
                int m = Math.max(tx, ty);
                int prev_ty = Math.max(tx, sy);
                int jump = (ty - sy - 1) / m + 1;
                if (tx == sx) jump = 1;
                ty -= m * jump;
                moves += jump;
            } else {
                // tx == ty
                if (tx == sx && ty == sy) return moves;
                return -1;
            }
        }
        return (tx == sx && ty == sy) ? moves : -1;
    }
}
# @lc code=end

File: 3609.minimum-moves-to-reach-target-in-grid.go
#
# @lc app=leetcode id=3609 lang=golang
#
# [3609] Minimum Moves to Reach Target in Grid
#
# @lc code=start
func minMoves(sx int, sy int, tx int, ty int) int {
    type Point struct { x, y int }
    queue := []Point{{sx, sy}}
    visited := make(map[Point]bool)
    visited[Point{sx, sy}] = true
    steps := 0
    for len(queue) > 0 {
        nextQueue := []Point{}
        for _, p := range queue {
            if p.x == tx && p.y == ty {
                return steps
            }
            m := max(p.x, p.y)
            newPoints := []Point{{p.x + m, p.y}, {p.x, p.y + m}}
            for _, np := range newPoints {
                if np.x <= tx && np.y <= ty && !visited[np] {
                    visited[np] = true
                    nextQueue = append(nextQueue, np)
                }
            }
        }
        queue = nextQueue
        steps++
    }
    return -1 // If we exhaust options without reaching (tx, ty), return -1. 
}
func max(a, b int) int { if a > b { return a }; return b } 
# @lc code=end